packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -1359,19 +1383,38 @@ class ParentDataElement extends _ProxyElement<ParentDataWidget> {
 class InheritedElement extends _ProxyElement<InheritedWidget> {
   InheritedElement(InheritedWidget widget) : super(widget);
 
+  final Set<Element> _dependants = new HashSet<Element>();
+
+  void _updateInheritance() {
+    if (_parent?._inheritedWidgets != null)
+      _inheritedWidgets = new Map<Type, InheritedElement>.from(_parent._inheritedWidgets);
+    else
+      _inheritedWidgets = new Map<Type, InheritedElement>();
+    _inheritedWidgets[widget.runtimeType] = this;
+  }
+
+  void debugDeactivated() {
+    assert(() {
+      assert(_dependants.isEmpty);
+      return true;
+    });
+    super.debugDeactivated();
+  }
+
   void notifyDescendants(InheritedWidget oldWidget) {
     if (!widget.updateShouldNotify(oldWidget))
       return;
     final Type ourRuntimeType = widget.runtimeType;
-    void notifyChildren(Element child) {
-      if (child._dependencies != null &&
-          child._dependencies.contains(ourRuntimeType)) {
-        child.dependenciesChanged(ourRuntimeType);
-      }
-      if (child.runtimeType != ourRuntimeType)
-        child.visitChildren(notifyChildren);
+    for (Element dependant in _dependants) {
+      dependant.dependenciesChanged(ourRuntimeType);
```

Comment:
There's exactly one custom or this entire notification system, `InkResponse`'s `State`, which does:
```dart
  void dependenciesChanged(Type affectedWidgetType) {
    if (affectedWidgetType == Theme && _lastHighlight != null)
      _lastHighlight.color = Theme.of(context).highlightColor;
  }
```